### PR TITLE
Giphy Block

### PR DESCRIPTION
**Pull request by jeffersonrabb**
#### Changes proposed in this Pull Request

The Giphy block is a straightforward integration that allows the user to explicitly specify a Gif by URL or search by keyword. The keyword search is likely to be the most common use. 

Note: the block resizing is based on the approach described in this article https://blog.theodo.fr/2018/01/responsive-iframes-css-trick/

#### Testing instructions

Jurassic Ninja
- https://jurassic.ninja/create?gutenpack&calypsobranch=try/giphy&branch=try/giphy
- Connect Jetpack
- Navigate to Settings->Jetpack Constants, check `JETPACK_BETA_BLOCKS`, and Save.

1) Add the block
2) Go to Giphy, find any Gif you like, copy the URL and paste into the block. The gif should show up.
3) Paste in a Giphy URL following this syntax: http://i.giphy.com/4ZFekt94LMhNK.gif. Past the URL into the block. The gif should show up.
4) Type any text in to the search field. DIfferent Gifs should show up for anything you search for. 
5) Verify that the text field appears when you click on the block. 
6) Verify that the text field disappears after 3.5 seconds of inactivity, unless no search text has been provided in which case it remains visible.
7) Verify caption behaves much like the core Image block.
8) Verify block alignment options all work as expected.

### CODE REVIEW

**Comment by jeffersonrabb** [comment]
> I can't doubleclick to highlight all the text in the input field. I tried to click rapidly on a keyword so that I can select it all at once and retype something else, but couldn't. It would be good to make that possible.

In my testing I'm able to double-click and select the entire field even if there are multiple words (see screen recording below) Could you give me some additional detail on your environment, or possibly a screen recording?

https://cloudup.com/cSeY0yoGxQZ

**Comment by jeffersonrabb** [comment]
> On smaller instances of the gif, the input field is too small and should widen a bit to show the placeholder text if possible.

Here's a screen capture of proposed new sizing of the text field, from wide to narrow browser widths. Before I commit the change, does this do the trick @mapk?

https://cloudup.com/cYFmVarEKWp

**Comment by mapk** [comment]
>Could you give me some additional detail on your environment, or possibly a screen recording?

I'm on a Macbook Pro using Chrome (latest release) and still having trouble double-clicking.
https://cloudup.com/cczGVqu9fUR

At the end, I resort to using the keyboard to select the word and delete it.

**Comment by jeffersonrabb** [comment]
> I'm on a Macbook Pro using Firefox (latest release) and still having trouble double-clicking.
> https://cloudup.com/cczGVqu9fUR

Looks like this is a Firefox issue that occurs when you have a text input within a button element. Changing the button to a plain div resolved it. Fixed in https://github.com/Automattic/wp-calypso/pull/29638/commits/17ec7511074bd4c7191c4a2e2ea68b7c75c6a7aa

**Comment by jeffersonrabb** [comment]
> Here's a screen capture of proposed new sizing of the text field, from wide to narrow browser widths.

Fixed in https://github.com/Automattic/wp-calypso/pull/29638/commits/f9532f45a405078251d37d0fd6c378e8743b7f27

**Comment by thomasguillot** [comment]
https://cloudup.com/cFBJq9RLp6R

The search input goes to the very left of the block when selecting different alignment options when it should always be centred.

**Comment by thomasguillot** [comment]
Are we considering offering more than 1 result? This seems quite restrictive especially compared to competitors (e.g. CoBlocks)

screenshot 2019-01-15 at 16 20 43

**Comment by thomasguillot** [comment]
Let's use a Giphy icon for the block. I created one based on Material: https://cldup.com/rfwqWsMR4u.svg

**Comment by thomasguillot** [comment]
The "Paste a Giphy URL" doesn't seem to work. I always get the same image of an iPhone. I get a better result with Core's Image block.

https://cloudup.com/cEKGSK2dTRh

**Comment by jeffersonrabb** [comment]
> Are we considering offering more than 1 result? This seems quite restrictive especially compared to competitors (e.g. CoBlocks)

Good question. The rationale for offering only one was based in simplicity, delight, and instant gratification. Type anything and moments later you have a visualization of it. You can also paste in a Giphy URL, so for a discerning user they could go to Giphy.com, find the ideal image using all of the advanced search functionality, and paste the URL in the block. 

That said, offering multiple options is certainly doable. If we added this feature, I would advocate for setting the main image immediately just like the status quo. However, a line of alternate thumbnails would also appear along the bottom of the image. Clicking any of these thumbnails would replace the main image. @scruffian, what's your opinion - do you think we should go for this feature here?

**Comment by jeffersonrabb** [comment]
> The "Paste a Giphy URL" doesn't seem to work. I always get the same image of an iPhone. I get a better result with Core's Image block.

@thomasguillot could you drop in the Giphy URL you had trouble with? The block supports a couple of different URL formats, but it's possible that there are additional ones that were missed.

**Comment by thomasguillot** [comment]
> I would advocate for setting the main image immediately just like the status quo. However, a line of alternate thumbnails would also appear along the bottom of the image. Clicking any of these thumbnails would replace the main image.

This

**Comment by thomasguillot** [comment]
> @thomasguillot could you drop in the Giphy URL you had trouble with? The block supports a couple of different URL formats, but it's possible that there are additional ones that were missed.

Yep: https://media.giphy.com/media/9Pmfi0HtSpGSiuB991/giphy.gif

I'm simply using the link icon

screenshot 2019-01-15 at 16 54 41

**Comment by jeffersonrabb** [comment]
> Giphy does not always display the most relevant clip and their search is sorta-kinda not that great.
> 
> Slack solved this by adding a Shuffle button like so:

☝️  is a good idea too, @aldavigdis. @scruffian when you take a look you might consider this and the thumbnails, and see what you think.

**Comment by scruffian** [comment]
@jeffersonrabb I think it depends on effort. If either option are going to take significant time/energy then I think its best to do it later. However if you can do it quickly without too much energy then either solution is good.

**Comment by simison** [comment]
Regardless, it seems like a good iteration to do in a follow up PR.

**Comment by aldavigdis** [comment]
I also wonder if the caption field is required per se.

There is also a `title` attribute is in the `iframe` tag, but I wonder if we can modify it to include something like `Giphy clip: Ron Swanson`, where `Ron Swanson` is the search term in question.

**Comment by jeffersonrabb** [comment]
> Let's use a Giphy icon for the block. I created one based on Material:

Icon replaced in https://github.com/Automattic/wp-calypso/pull/29638/commits/5e2c846af055d60ab0d7ac384181aff346eeca4a

**Comment by jeffersonrabb** [comment]
> The "Paste a Giphy URL" doesn't seem to work. I always get the same image of an iPhone. I get a better result with Core's Image block.

This should be fixed as of https://github.com/Automattic/wp-calypso/pull/29638/commits/6760bd0632e9074745970cf5ff1f881e7cb516c5. The block now supports these three URL formats:

https://media.giphy.com/media/gt0hYzKlMpfOg/giphy.gif
https://giphy.com/embed/gt0hYzKlMpfOg
http://i.giphy.com/gt0hYzKlMpfOg.gif

**Comment by jeffersonrabb** [comment]
> The search input goes to the very left of the block when selecting different alignment options when it should always be centred.

This is resolved in https://github.com/Automattic/wp-calypso/pull/29638/commits/d39a8727e86fd555f052af27e8935921259c4977

**Comment by sirreal** [comment]
Rebased and pushed a few small changes:

- Drop vendor prefixes from styles (autoprefixer handles)
- Fix irregular indentation

I'll push a few more small changes and give a more extensive review soon.

File: client/gutenberg/extensions/giphy/settings.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		},
+		topPadding: {
+			type: 'number',
+			default: 56.2,
+		},
+	},
+	supports: {
+		html: false,
+	},
+	apiKey: 'LtJmY9DnDJUA0',
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/settings.js]: This seems problematic.

**Comment by thomasguillot** [comment]
> According to some of the docs, it seems that we'll also need to include Giphy branding in the block in the editor, shall we get some design input there?

If we really need to do that then I would suggest to use a standard Gutenberg placeholder when adding the block instead of the current iframe placeholder.

screenshot 2019-01-22 at 10 13 04

I converted the "Powered by GIPHY" logo to an SVG so it's easier to include: https://cloudup.com/cu94cGM8s4I

File: client/gutenberg/extensions/giphy/edit.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+			return this.fetch( this.urlForId( giphyID ) );
+		}
+
+		return this.fetch( this.urlForSearch( searchText ) );
+	}, 250 );
+
+	urlForSearch = searchText => {
+		const { apiKey } = settings;
+		return `//api.giphy.com/v1/gifs/search?q=${ encodeURIComponent(
+			searchText
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/edit.js]: I replaced the ` ` to `+` encoding here with full `encodeURIComponent encoding.

You could see that typing `nope&offset=1` would be passed through cleanly into the query, resulting in a different offset rather than searching for the term.

Likely, all of these query parts should be encoded: `apiKey`, `giphyId`. They can come from attributes in the post and may not be clean.

File: client/gutenberg/extensions/giphy/edit.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		const { searchText } = attributes;
+		return searchText && searchText.length > 0;
+	};
+
+	render() {
+		const { attributes, className, isSelected, setAttributes } = this.props;
+		const { align, caption, giphyUrl, searchText, topPadding } = attributes;
+		const { focus } = this.state;
+		const style = {
+			paddingTop: `${ topPadding }%`,
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/edit.js]: It would be nice to just use the attribute without appending `%`, seems like we could store that complete string.

File: client/gutenberg/extensions/giphy/settings.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		giphyUrl: {
+			type: 'string',
+			default: '//giphy.com/embed/ZgTR3UQ9XAWDvqy9jv',
+		},
+		searchText: {
+			type: 'string',
+		},
+		topPadding: {
+			type: 'number',
+			default: 56.2,
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/settings.js]: 2 thoughts here:

- Just store the percent string we need in the style? (need to change the type)
- Did you attempt to parse this from the attribute? We're duplicating the ways this is stored. That's fine if it's complicated or fragile to get the value it out of the style attribute.

```suggestion
			default: '56.2%',
```

File: client/gutenberg/extensions/giphy/edit.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+				</BlockControls>
+				<div className={ classes }>
+					<figure style={ style }>
+						<div
+							className="wp-block-jetpack-giphy_cover"
+							onClick={ this.setFocus }
+							onKeyDown={ this.setFocus }
+							ref={ this.textControlRef }
+							role="button"
+							tabIndex="0"
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/edit.js]: This wasn't working, it was `tabindex` rather than `tabIndex` as React expected.

However, I'm not sure that this should be tab-focusable. We can tab to the input itself, which seems preferable.

Should we drop this?

File: client/gutenberg/extensions/giphy/edit.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+						onChange={ value => setAttributes( { align: value } ) }
+						value={ align }
+					/>
+				</BlockControls>
+				<div className={ classes }>
+					<figure style={ style }>
+						<div
+							className="wp-block-jetpack-giphy_cover"
+							onClick={ this.setFocus }
+							onKeyDown={ this.setFocus }
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/edit.js]: Can you explain why this is necessary?

**Comment by sirreal** [comment]
> If we really need to do that then I would suggest to use a standard Gutenberg placeholder when adding the block instead of the current iframe placeholder.

@thomasguillot I like that and it seems to follow Gutenberg patterns nicely. What about when searching on a Giphy block with an images that's loaded. We'd need to show branding there as well.

**Comment by thomasguillot** [comment]
> What about when searching on a Giphy block with an images that's loaded. We'd need to show branding there as well.

@sirreal if it's really required even at this point (?), then we could add a very subtle branding:

screenshot 2019-01-22 at 10 56 01

Another idea would be to simply have the branding in the sidebar making the branding less invasive:

screenshot 2019-01-22 at 10 59 54

File: client/gutenberg/extensions/giphy/edit.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+					</div>
+					<iframe src={ giphyUrl } title={ searchText } />
+				</figure>
+				{ ( ! RichText.isEmpty( caption ) || isSelected ) && (
+					<RichText
+						className="caption"
+						inlineToolbar
+						onChange={ value => setAttributes( { caption: value } ) }
+						placeholder={ __( 'Write caption…' ) }
+						tagName="figcaption"
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/edit.js]: `figcaption` is only permitted as the first or last direct descendent of `figure`:

> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figcaption

We may need to restructure this a bit.

File: client/gutenberg/extensions/giphy/index.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ * Internal dependencies
+ */
+import { settings as giphySettings } from './settings.js';
+import edit from './edit';
+import './editor.scss';
+import './style.scss';
+
+export const { name } = giphySettings;
+
+export const settings = {
```
Comment (sirreal) [inline client/gutenberg/extensions/giphy/index.js]: This seems pretty unusual. Shall we export `name` and `settings` directly here as we do in other blocks?

**Comment by sirreal** [comment]
I've rebased and left a lot of feedback.

I've also pushed several changes. The most significant change I've made was to simplify the alignment by declaring `supports: { align: true }`, which allows us to simplify this block and leverage Gutenberg's built-in functionality.

**Comment by simison** [comment]
>What's the story with the API keys

Would be good to know up front how much work this might be and if it's different for wpcom vs self-hosted Jetpack.

With Maps block it was clear that we couldn't ship our API key with the block, but is there a difference here in terms of user expectations for the block "to just work" or how Giphy thinks about multiple users using the same API key?

**Comment by simison** [comment]
>Another idea would be to simply have the branding in the sidebar making the branding less invasive:
>
> screenshot 2019-01-22 at 10 59 54

Gutenberg design docs mention keeping branding away from the description:

image

https://wordpress.org/gutenberg/handbook/designers-developers/designers/block-design/#block-description

...but maybe this is ok:

image

File: client/gutenberg/extensions/giphy/settings.js (hunk 1 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/settings.js]: The API key in use here is the one that was used in the Atavist block in production for the past ~4 years. I'm 90% sure it's a production key, although it's immaterial since we will need an A8C key anyway (which I'd forgotten, so thank you!). I'll contact Giphy today about getting one. Unlike Map block, the Giphy key is unrestricted and has no costs associated with it. Given this,  what's your hesitation around committing the key as a setting? The key is visible in the XHR requests anyway, and key theft isn't a concern the way it is for a monetized service.

**Comment by jeffersonrabb** [comment]
In the Atavist version, the Giphy branding is presented to the right of the caption, whether there is one or not. Unlike sidebar or placeholder placements, this is always visible in both the editor and view environments. @thomasguillot what do you think of going this way? Here are screenshots for Atavist block in editor and view:

atavist giphy in editor
atavist giphy in view

File: client/gutenberg/extensions/giphy/edit.js (hunk 5 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: The `tabIndex` was added in response to an ESLint error:

`Elements with the 'button' interactive role must be tabbable  jsx-a11y/interactive-supports-focus`

File: client/gutenberg/extensions/giphy/settings.js (hunk 1 above)
Comment (sirreal) [reply client/gutenberg/extensions/giphy/settings.js]: 👍 No worries if it's save to use and share publicly and you're on top of it 🙂

File: client/gutenberg/extensions/giphy/edit.js (hunk 6 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: In order to present the selected gif unobscured, the text input disappears after 3.5s of inactivity. Clicking on the block makes it visible again, and sets keyboard focus to the input, which explains the `onClick` handler. The `onKeyDown` handler is present for this similar scenario:

- User starts typing, pauses for 4 seconds to consider exact what animated gif they want to show the world
- Because of timeout, the input disappears
- When they start typing again, input reappears, and the field regains focus.

File: client/gutenberg/extensions/giphy/settings.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	description: __( 'Add a giphy graphic.' ),
+	attributes: {
+		align: {
+			type: 'string',
+			default: 'center',
+		},
+		caption: {
+			type: 'string',
+		},
+		giphyUrl: {
```
[the remark this answers is not in the document]
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/settings.js]: I have no problem doing this or the requested padding percentage change, but they seem to be subtly contradictory. In the padding case, you're requesting a change that increases the length of the stored attribute so that it can be used as-is without any programmatic modification. In this case the suggestion seems to be the opposite—store the shorter, essential value and transform programmatically when the value is needed for rendering. Again, a very subtle point and I'm happy to make both changes as requested :-), I'm just trying to grasp the full nature of the recommendation.

File: client/gutenberg/extensions/giphy/settings.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
+import { Path, SVG } from '@wordpress/components';
+
+export const settings = {
+	name: 'giphy',
+	icon: (
+		<SVG xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">
+			<Path fill="none" d="M0 0h24v24H0V0z" />
```
[the remark this answers is not in the document]
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/settings.js]: Question for @thomasguillot - is something missing from the Icon? Something you said earlier made me expect something specifically Gif-related about the icon, and in its current state I don't see anything like that. This + @sirreal's question about the empty `Path` made me wonder if something went missing or got corrupted? 

screen shot 2019-01-22 at 7 14 11 am

**Comment by thomasguillot** [comment]
> In the Atavist version, the Giphy branding is presented to the right of the caption, whether there is one or not. Unlike sidebar or placeholder placements, this is always visible in both the editor and view environments.

Yeah I think that would be fine. I don't think we need the branding on the front-end since when we hover we get the Giphy branding.

File: client/gutenberg/extensions/giphy/settings.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (thomasguillot) [reply client/gutenberg/extensions/giphy/settings.js]: All material icons have an invisible path to make sure they stick to the 24x24 grid. If we remove it the icon will not follow that pattern.

File: client/gutenberg/extensions/giphy/settings.js (hunk 1 above)
Comment (simison) [reply client/gutenberg/extensions/giphy/settings.js]: Perfect, thanks for elaborating, Jeff!

File: client/gutenberg/extensions/giphy/edit.js (hunk 6 above)
Comment (sirreal) [reply client/gutenberg/extensions/giphy/edit.js]: I overstepped in ae6ae7cd2275b1fa880830d658562ae06bcbb40c by making the focused input always visible.

I'll roll that back and this makes sense again 👍

File: client/gutenberg/extensions/giphy/settings.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (sirreal) [reply client/gutenberg/extensions/giphy/settings.js]: That's a good point and thanks for mentioning it here

🤔 💭 

After much careful consideration 🤕, let's store the padding as a complete percent string and forget about this change, storing the full giphy URL is good 👍

File: client/gutenberg/extensions/giphy/settings.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (sirreal) [reply client/gutenberg/extensions/giphy/settings.js]: The extra `Path` doesn't hurt anything, but in the context of a unique SVG document containing a single logo with a well dimensioned `viewBox` it's probably not necessary, unless I'm missing something.

Pretty minor, shouldn't make much difference either way.

File: client/gutenberg/extensions/giphy/edit.js [hunk 11]
```diff
@@ -16,15 +16,15 @@ import { debounce } from 'lodash';
...
-	constructor() {
-		super( ...arguments );
-		this.state = {
-			focus: false,
-		};
-		this.timer = null;
-		this.textControlRef = createRef();
-		this.debouncedParseSearch = debounce( this.parseSearch, 250 );
-	}
+	timer = null;
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/gutenberg/extensions/giphy/edit.js]: They do, but there's a late-stage proposal (shipping to browsers imminently) to add ["field declarations"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Field_declarations) to classes. We've had this enabled for quite a while in Calypso and it's an established pattern you'll tend to see in the code base:

For React component classes, initial state should be in the `state` property on the instance upon instantiation, and field properties handles this nice and succinctly.

tl;dr: They're equivalent and this is a stylistic change.

<details>

<summary>funsies</summary>

This works in Chrome 72 (currently on the Chrome canary release channel)

```js
class X {
  static staticField = '😎';
  instanceField = '😵';
  constructor( builtWith ) {
    this.builtWith = builtWith;
  }
}

const y = new X( '👋' );
y // {instanceField: "😵", builtWith: "👋"}
y.constructor.staticField // "😎"
X.staticField // "😎"
```

</details>

[the remark this answers is not in the document]
Comment (simison) [reply client/gutenberg/extensions/giphy/edit.js]: I'd say the key aspect of this change is that it makes the code consistent with the rest of Calypso codebase, while previously the component was consistent with core Gutenberg. 
That was okay too, but better be consistent across one codebase where the code lives.

[the remark this answers is not in the document]
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: Sounds good, thanks for the explanation!

File: client/gutenberg/extensions/giphy/edit.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+	};
+
+	splitAndLast = ( array, delimiter ) => {
+		const split = array.split( delimiter );
+		return split[ split.length - 1 ];
+	};
+
+	fetch = url => {
+		const { setAttributes } = this.props;
+		const xhr = new XMLHttpRequest();
```
[the remark this answers is not in the document]
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: Did a quick test with `apiFetch` and encountered a CORS error. In the test I replaced the `fetch()` function with the following:

```fetch = url => {
    const { setAttributes } = this.props;
    apiFetch( { url } ).then(
      result => {
        console.log( 'API Success', result );
      },
      result => {
        console.log( 'API Failure', result );
      }
    );
};```

This resulted in the following error in the console:

```Access to fetch at 'http://api.giphy.com/v1/gifs/search?q=ron%20swanson&api_key=LtJmY9DnDJUA0&limit=1&_locale=user' from origin 'http://localhost' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'.```

Also noting that the Gutenberg Handbook's API Fetch docs ( https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-api-fetch/ ) describe it as `Utility to make WordPress REST API requests.`. If we can make it work, I think we should, but it may be that the library wasn't intended to be used with external APIs.

[the remark this answers is not in the document]
Comment (simison) [reply client/gutenberg/extensions/giphy/edit.js]: Interesting — I've understood `wp.apiFetch` to be just a thin layer over [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and kind of generic utility.

File: client/gutenberg/extensions/giphy/edit.js (hunk 6 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: Rolled https://github.com/Automattic/wp-calypso/commit/ae6ae7cd2275b1fa880830d658562ae06bcbb40c back in https://github.com/Automattic/wp-calypso/pull/29638/commits/24f732045499a6f596b8e24cbaa1de5ccbe99249. All fixed!

File: client/gutenberg/extensions/giphy/edit.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (simison) [reply client/gutenberg/extensions/giphy/edit.js]: Asked: https://wordpress.slack.com/archives/C02RQBWTW/p1548167770493500

[the remark this answers is not in the document]
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: I encountered the same problem with the Mapbox location lookups in the Map block and ended up choosing the path of least resistance and using vanilla XMLHttpRequest: https://github.com/Automattic/wp-calypso/blob/master/client/gutenberg/extensions/map/location-search/index.js#L65-L79

If we can find resolution that would be great as using `apiFetch` is much nicer than rolling our own every time we need to communicate with a non-WP API.

File: client/gutenberg/extensions/giphy/edit.js (hunk 3 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: Done in https://github.com/Automattic/wp-calypso/pull/29638/commits/aefd179d872ea0c58e0de15d435bd68fa65e6167. Note that this commit corresponds to https://github.com/Automattic/jetpack/pull/10989/commits/181de07ce813cf0cad08be143c41f33eac8b39e4, and requires pulling current code from Jetpack repo so that Preview/Publish continues to work properly.

File: client/gutenberg/extensions/giphy/settings.js (hunk 4 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/settings.js]: Done in https://github.com/Automattic/wp-calypso/commit/aefd179d872ea0c58e0de15d435bd68fa65e6167 and https://github.com/Automattic/jetpack/commit/181de07ce813cf0cad08be143c41f33eac8b39e4

File: client/gutenberg/extensions/giphy/edit.js (hunk 2 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: All are encoded as of https://github.com/Automattic/wp-calypso/pull/29638/commits/d077f8ce19d1073f02d9b7fa85883d3f3425431b

File: client/gutenberg/extensions/giphy/edit.js (hunk 7 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/edit.js]: Before addressing this, perhaps we should consider removing the caption as @aldavigdis asks here: https://github.com/Automattic/wp-calypso/pull/29638#issuecomment-454818106. @scruffian what do you think? Caption or no caption? The Atavist block does have a caption, although I'll be able to work around this in migrations by converting an Atavist Giphy block to Gutenberg Giphy + Gutenberg Paragraph (or Heading).

File: client/gutenberg/extensions/giphy/index.js (hunk 8 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/index.js]: To confirm, are you asking why this block has a separate `settings.js` file at all as opposed to including all settings right in `index.js`?

Comment (sirreal) [reply client/gutenberg/extensions/giphy/index.js]: Yep! Most other Jetpack blocks have an index which exports `name` and `settings`, which are defined there:

https://github.com/Automattic/wp-calypso/blob/8a2f6265eb9381a1fabdbe7b01d8253bae3dc506/client/gutenberg/extensions/tiled-gallery/index.js#L91
https://github.com/Automattic/wp-calypso/blob/8a2f6265eb9381a1fabdbe7b01d8253bae3dc506/client/gutenberg/extensions/tiled-gallery/index.js#L102

https://github.com/Automattic/wp-calypso/blob/8a2f6265eb9381a1fabdbe7b01d8253bae3dc506/client/gutenberg/extensions/markdown/index.js#L17-L19

Is the indirection and re-export of parts of the object exported from settings here providing value, or can we define the settings right here? I see that the `apiKey` is shared, but that might be better to move into a shared constants file.

File: client/gutenberg/extensions/giphy/edit.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (sirreal) [reply client/gutenberg/extensions/giphy/edit.js]: `apiFetch` as a wrapper over `window.fetch` should work fine. I'm seeing `window.fetch` requests working without any CORS issues.

I do see some issues with `apiFetch` related to some middleware it's using that is preventing external APIs —not `public-api.wordpress.com`— from being reached.

This doesn't need to block this from landing, especially as a beta block. We can make this change after it's landed and the requirements are satisfied.

[the remark this answers is not in the document]
Comment (sirreal) [reply client/gutenberg/extensions/giphy/edit.js]: This patch mostly works using  `window.apiFetch`, `@wordpress/api-fetch` should be a bit nicer once we get over the hurdles:

```patch
diff --git a/client/gutenberg/extensions/giphy/edit.js b/client/gutenberg/extensions/giphy/edit.js
index 93bab39d28..1877a7c8d6 100644
--- a/client/gutenberg/extensions/giphy/edit.js
+++ b/client/gutenberg/extensions/giphy/edit.js
@@ -1,8 +1,7 @@
-/** @format */
-
 /**
  * External dependencies
  */
+import apiFetch from '@wordpress/api-fetch';
 import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
 import classNames from 'classnames';
 import { Component, createRef } from '@wordpress/element';
@@ -77,31 +76,29 @@ class GiphyEdit extends Component {
 		return split[ split.length - 1 ];
 	};
 
-	fetch = url => {
-		const { setAttributes } = this.props;
-		const xhr = new XMLHttpRequest();
-		xhr.open( 'GET', url );
-		xhr.onload = () => {
-			if ( xhr.status === 200 ) {
-				const res = JSON.parse( xhr.responseText );
-				const giphyData = res.data.length > 0 ? res.data[ 0 ] : res.data;
-				// No results
-				if ( ! giphyData.images ) {
-					return;
-				}
-				const calculatedPaddingTop = Math.floor(
-					( giphyData.images.original.height / giphyData.images.original.width ) * 100
-				);
-				const paddingTop = `${ calculatedPaddingTop }%`;
-				const giphyUrl = giphyData.embed_url;
-				setAttributes( { giphyUrl, paddingTop } );
-				this.maintainFocus( 500 );
-			} else {
-				// Error handling TK
-			}
-		};
-		xhr.send();
-	};
+	fetch = url =>
+		window
+			.fetch( url )
+			.then( resp => resp.json() )
+			.then(
+				( { data } ) => {
+					const { setAttributes } = this.props;
+					// No results
+					const giphyData = data.length > 0 ? data[ 0 ] : data;
+					if ( ! giphyData.images ) {
+						return;
+					}
+					const calculatedPaddingTop = Math.floor(
+						( giphyData.images.original.height / giphyData.images.original.width ) * 100
+					);
+					const paddingTop = `${ calculatedPaddingTop }%`;
+					const giphyUrl = giphyData.embed_url;
+					setAttributes( { giphyUrl, paddingTop } );
+					this.maintainFocus( 500 );
+				},
+				// eslint-disable-next-line no-console
+				'production' === process.env.NODE_ENV ? err => console.error( err ) : undefined
+			);
 
 	setFocus = () => {
 		this.maintainFocus();
```

[the remark this answers is not in the document]
Comment (sirreal) [reply client/gutenberg/extensions/giphy/edit.js]: I made the apiFetch change in #30332 and #30333. Depending on how reviews unfold that will give us some flexibility on how we land things.

This isn't a blocker for this PR.

**Comment by sirreal** [comment]
To be clear, there remain things to do, but they can be handled as follow-up. In particular, proper Giphy attribution according to their guidelines.

**Comment by scruffian** [comment]
> To be clear, there remain things to do, but they can be handled as follow-up. In particular, proper Giphy attribution according to their guidelines.

Is there anything else outstanding?

**Comment by sirreal** [comment]
> Is there anything else outstanding?

I'll documenting all existing concerns as issues. I'm working through them now.

**Comment by sirreal** [comment]
Relevant follow-ups:

- https://github.com/Automattic/wp-calypso/issues/30343
- https://github.com/Automattic/wp-calypso/issues/30345
- https://github.com/Automattic/wp-calypso/issues/30346
- https://github.com/Automattic/wp-calypso/issues/30348

File: client/gutenberg/extensions/giphy/settings.js (hunk 1 above)
Comment (jeffersonrabb) [reply client/gutenberg/extensions/giphy/settings.js]: Giphy has approved the API for production use, so we can call this resolved.
